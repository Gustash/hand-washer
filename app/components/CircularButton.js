import React from 'react'
import { TouchableOpacity, View, StyleSheet, Text } from 'react-native'
import { colors } from 'app/config'

export const DEFAULT_CIRCULAR_BUTTON_SIZE = 204

const CircularButton = ({
  size = DEFAULT_CIRCULAR_BUTTON_SIZE,
  onPress,
  text,
  icon,
  style,
}) => (
  <TouchableOpacity
    style={[
      styles.container,
      { width: size, height: size, borderRadius: size / 2 },
      style,
    ]}
    onPress={onPress}
    disabled={!onPress}
  >
    <View style={styles.icon}>{icon}</View>
    <Text style={styles.text}>{text}</Text>
  </TouchableOpacity>
)

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.lightBlue,
    shadowColor: colors.black,
    shadowOpacity: 0.2,
    shadowOffset: {
      height: 2,
    },
    shadowRadius: 16,
    padding: 30,

    elevation: 16,
  },
  icon: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  text: {
    color: colors.white,
    textAlign: 'center',
    fontSize: 21,
  },
})

export default CircularButton
