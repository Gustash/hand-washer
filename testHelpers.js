export const findNodeWithTestId = (wrapper, testId) =>
  wrapper.findWhere(node => node.prop('testId') === testId)
