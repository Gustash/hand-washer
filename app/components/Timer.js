import React, { useState, useRef, useEffect } from 'react'
import Animated, {
  Easing,
  useCode,
  call,
  stopClock,
  timing,
  startClock,
  cond,
  clockRunning,
  and,
  not,
} from 'react-native-reanimated'
import { StyleSheet, Text, View } from 'react-native'
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5'

import CircularButton, { DEFAULT_CIRCULAR_BUTTON_SIZE } from './CircularButton'
import CircleProgress from './CircleProgress'
import { colors } from 'app/config'

const { Value, Clock } = Animated

const Timer = ({
  duration,
  strokeWidth = 9,
  size = DEFAULT_CIRCULAR_BUTTON_SIZE + strokeWidth * 2,
  style,
  onTimerFinished,
}) => {
  const [currentProgress, setCurrentProgress] = useState(duration / 1000)
  const [timerFinished, setTimerFinished] = useState(false)

  const clock = useRef(new Clock()).current
  const progress = useRef(new Value(0)).current
  const finished = useRef(new Value(0)).current
  const frameTime = useRef(new Value(0)).current
  const time = useRef(new Value(0)).current

  useCode(() => {
    const clockIsRunning = clockRunning(clock)

    return [
      timing(
        clock,
        { finished, position: progress, frameTime, time },
        { toValue: 1, duration, easing: Easing.linear }
      ),
      cond(and(finished, clockIsRunning), stopClock(clock)),
      cond(and(not(finished), not(clockIsRunning)), startClock(clock)),
      call([clockIsRunning, finished], ([running, clockFinished]) => {
        if (clockFinished && running) {
          setTimerFinished(true)
        }
      }),
      call([frameTime], value => {
        if (clockIsRunning) {
          setCurrentProgress(Math.round((duration - value) / 1000))
        }
      }),
    ]
  }, [progress, duration, clock, finished, frameTime, time, onTimerFinished])

  useEffect(() => {
    if (timerFinished) {
      onTimerFinished?.()
    }
  }, [timerFinished, onTimerFinished])

  // noinspection JSSuspiciousNameCombination
  return (
    <View style={style}>
      <CircleProgress
        testId="timer-circle-progress"
        size={size}
        progress={progress}
        strokeWidth={strokeWidth}
      />
      <CircularButton
        testId="timer-circular-button"
        icon={
          timerFinished ? (
            <FontAwesome5
              name="thumbs-up"
              size={48}
              color={colors.white}
              solid
            />
          ) : (
            <Text style={styles.timerText}>{currentProgress}</Text>
          )
        }
        text={timerFinished ? 'Well Done' : 'Seconds'}
        style={[
          styles.timer,
          {
            top: strokeWidth,
            left: strokeWidth,
          },
        ]}
        size={size - strokeWidth * 2}
      />
    </View>
  )
}

const styles = StyleSheet.create({
  timerText: {
    fontSize: 48,
    color: colors.white,
  },
  timer: {
    position: 'absolute',
  },
})

export default Timer
