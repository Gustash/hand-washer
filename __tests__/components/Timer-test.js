import React from 'react'
import { Text } from 'react-native'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5'
import { useCode, call, clockRunning } from 'react-native-reanimated'

import Timer from '../../app/components/Timer'
import CircularButton from '../../app/components/CircularButton'
import { findNodeWithTestId } from '../../testHelpers'

const mockProps = props => ({
  duration: 20000,
  style: { marginTop: 1000 },
  onTimerFinished: jest.fn(),
  ...props,
})

describe('<Timer />', () => {
  let props
  let wrapper

  beforeEach(() => {
    useCode.mockReset()

    props = mockProps()
    wrapper = shallow(<Timer {...props} />)
  })

  it('matches snapshot', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('initializes bottom text', () => {
    expect(
      wrapper.containsMatchingElement(<CircularButton text="Seconds" />)
    ).toBe(true)
  })

  it('initializes second count text', () => {
    const Icon = () =>
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('icon')
    let icon = shallow(<Icon />)

    expect(icon).toMatchElement(
      <Text
        style={{
          fontSize: 48,
          color: '#FFFFFF',
        }}
      >
        20
      </Text>
    )
  })

  it('defaults stroke width', () => {
    expect(
      findNodeWithTestId(wrapper, 'timer-circle-progress').prop('strokeWidth')
    ).toBe(9)
  })

  it('defaults size', () => {
    expect(
      findNodeWithTestId(wrapper, 'timer-circle-progress').prop('size')
    ).toBe(222)
  })

  it('applies stroke width', () => {
    wrapper.setProps({
      ...props,
      strokeWidth: 16,
    })

    expect(
      findNodeWithTestId(wrapper, 'timer-circle-progress').prop('strokeWidth')
    ).toBe(16)
  })

  it('applies size', () => {
    wrapper.setProps({
      ...props,
      size: 16,
    })

    expect(
      findNodeWithTestId(wrapper, 'timer-circle-progress').prop('size')
    ).toBe(16)
  })

  it('applies style', () => {
    expect(wrapper.first().prop('style')).toEqual({ marginTop: 1000 })

    wrapper.setProps({
      ...props,
      style: { marginBottom: 1000 },
    })

    expect(wrapper.first().prop('style')).toEqual({ marginBottom: 1000 })
  })

  it('calls useCode', () => {
    expect(useCode).toHaveBeenCalled()
  })

  it('sets current progress', () => {
    useCode.mockImplementationOnce(a => a())
    call.mockImplementationOnce((a, b) => b([1, 0]))
    call.mockImplementationOnce((a, b) => b(2000))
    clockRunning.mockReturnValue(true)

    wrapper = shallow(<Timer {...props} />)

    const Icon = () =>
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('icon')
    let icon = shallow(<Icon />)

    expect(icon).toMatchElement(
      <Text
        style={{
          fontSize: 48,
          color: '#FFFFFF',
        }}
      >
        18
      </Text>
    )
  })

  it('keeps current progress when clock not running', () => {
    useCode.mockImplementationOnce(a => a())
    call.mockImplementationOnce((a, b) => b([0, 1]))
    call.mockImplementationOnce((a, b) => b(21000))
    clockRunning.mockReturnValue(false)

    wrapper = shallow(<Timer {...props} />)

    const Icon = () =>
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('icon')
    let icon = shallow(<Icon />)

    expect(icon).toMatchElement(
      <Text
        style={{
          fontSize: 48,
          color: '#FFFFFF',
        }}
      >
        20
      </Text>
    )
  })

  it('sets timer finished', () => {
    useCode.mockImplementationOnce(a => a())
    call.mockImplementationOnce((a, b) => b([1, 1]))

    wrapper = shallow(<Timer {...props} />)

    const Icon = () =>
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('icon')
    const icon = shallow(<Icon />)

    expect(
      icon.containsMatchingElement(
        <FontAwesome5 name="thumbs-up" size={48} color="#FFFFFF" solid />
      )
    ).toBe(true)

    expect(
      wrapper.containsMatchingElement(<CircularButton text="Well Done" />)
    ).toBe(true)
  })

  it('calls onTimerFinished when animation ends', () => {
    useCode.mockImplementationOnce(a => a())
    call.mockImplementationOnce((a, b) => b([0, 0]))

    shallow(<Timer {...props} />)

    useCode.mockImplementationOnce(a => a())
    call.mockImplementationOnce((a, b) => b([0, 1]))

    shallow(<Timer {...props} />)

    useCode.mockImplementationOnce(a => a())
    call.mockImplementationOnce((a, b) => b([1, 0]))

    shallow(<Timer {...props} />)

    expect(props.onTimerFinished).not.toHaveBeenCalled()

    useCode.mockImplementationOnce(a => a())
    call.mockImplementationOnce((a, b) => b([1, 1]))

    shallow(<Timer {...props} />)

    expect(props.onTimerFinished).toHaveBeenCalledTimes(1)
  })

  it('positions circular button properly regardless of stroke width', () => {
    expect(
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('style')
    ).toContainEqual({
      top: 9,
      left: 9,
    })
    expect(
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('size')
    ).toBe(204)

    wrapper.setProps({
      ...props,
      size: 400,
      strokeWidth: 100,
    })

    expect(
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('style')
    ).toContainEqual({
      top: 100,
      left: 100,
    })
    expect(
      findNodeWithTestId(wrapper, 'timer-circular-button').prop('size')
    ).toBe(200)
  })
})
