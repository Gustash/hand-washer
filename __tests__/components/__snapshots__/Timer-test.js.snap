// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> matches snapshot 1`] = `
<View
  style={
    Object {
      "marginTop": 1000,
    }
  }
>
  <CircleProgress
    progress={
      AnimatedValue {
        " __value": 0,
      }
    }
    size={222}
    strokeWidth={9}
    testId="timer-circle-progress"
  />
  <CircularButton
    icon={
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 48,
          }
        }
      >
        20
      </Text>
    }
    size={204}
    style={
      Array [
        Object {
          "position": "absolute",
        },
        Object {
          "left": 9,
          "top": 9,
        },
      ]
    }
    testId="timer-circular-button"
    text="Seconds"
  />
</View>
`;
